[clang] remove IsDefaulted bit from TemplateArgument

mizvekov opened the pull request:
The IsDefaulted bit being part of a canonical TemplateArgument doesn't make sense, as that information is not information a canonical type should have.

In C++, all template specialization types for ther same template are the same if the full list of template arguments is the same, an argument being defaulted or not doesn't matter.

Moreover, this information is already available in the sugared template specialization type, in the sense that, taking the as-written list and matching it up to the template parameters, any parameters which are left without a corresponding template argument must have been defaulted.

This patch besides removing that bit, changes the current DebugInfo users to derive that information from the as-written argument list. And it goes a little beyond that by wiring up the actual sugared TemplateArguments, so the Debug Info produced is also richer.

This patch is a performance improvement, as the TemplateArgument is one of the hottest data structures for C++ compilation:
<img width="1459" height="21" alt="image" src="https://github.com/user-attachments/assets/8c7afb31-c821-4c01-b07e-31974af75688" />

The small regression on `-O0 -g` test is explained by the increased amount of debug info generated.

mizvekov committed:
[clang] remove isDefaulted bit from TemplateArgument

The IsDefaulted bit being part of a canonical TemplateArgument
doesn't make sense, as that information is not information a
canonical type should have.

In C++, all template specialization types for ther same template
are the same if the full list of template arguments is the same,
an argument being defaulted or not doesn't matter.

Moreover, this information is already available in the sugared
template specialization type, in the sense that, taking the as-written
list and matching it up to the template parameters, any parameters which
are left without a corresponding template argument must have been defaulted.

This patch besides removing that bit, changes the current DebugInfo users to derive
that information from the as-written argument list. And it goes a little beyond
that by wiring up the actual sugared TemplateArguments, so the Debug Info
produced is also richer.

This patch is a performance improvement, as the TemplateArgument is one of the
hottest data structures for C++ compilation:

The small regression on `-O0 -g` test is explained by the increased amount of
debug info generated.

zwuis commented on clang/test/DebugInfo/CXX/template.cpp:
```diff
@@ -95,24 +95,25 @@ TC
+// CHECK: [[FOO_MEM]] = !DIDerivedType(tag: DW_TAG_typedef, name: "foo_mem", {{.*}}baseType: [[MEMINTPTR]])
   nullptr,
 //
 // In some future iteration we could possibly emit the value of a null member
 // function pointer as '{ i64, i64 } zeroinitializer' as it may be handled
 // naturally from the LLVM CodeGen side once we decide how to handle non-null
 // member function pointers. For now, it's simpler just to emit the 'i8 0'.
```
Should these be updated?

mizvekov replied:
I think this may still be a problem, doesn't just show up in this test anymore because we use the as-written template argument expression, but it probably does still happen for defaulted template arguments, as that still should use the converted template argument.

So I will probably just update that comment to mention that.

shafik commented:
> This patch doesn't touch LLDB. If LLDB uses this bit, then LLDB also needs to be updated in the same manner as we are updating DebugInfo here.

If the change breaks LLDB then you should also fix that break as well.

mizvekov commented:
> If the change breaks LLDB then you should also fix that break as well.

Yes of course, but the parts already in the PR can be reviewed independently of that.

shafik commented:
> > If the change breaks LLDB then you should also fix that break as well.
> 
> Yes of course, but the parts already in the PR can be reviewed independently of that.

Yeah, I think I would prefer if the fix for LLDB would be rolled up into the same PR to avoid issues w/ reverts/reapplies etc 

For sure the rest of the changes can be reviewed while you work on that.

steakhal commented:
I noticed that some tests are dropped in the PR, esp. eg. in the AST unittests.
Did you check the test coverage if that wouldn't shrink due to those removals?
In the ASTImporter I remember a few years ago we added some tests to exercise the uncovered ctors and getters or some AST constructs. If the coverage is as good as before, I'm good with the change. Thank you!

mizvekov commented:
> I noticed that some tests are dropped in the PR, esp. eg. in the AST unittests. Did you check the test coverage if that wouldn't shrink due to those removals? In the ASTImporter I remember a few years ago we added some tests to exercise the uncovered ctors and getters or some AST constructs. If the coverage is as good as before, I'm good with the change. Thank you!

The tests and parts of tests I removed were specifically testing for the removed bit, so I don't think any of this reduces test coverage.

Michael137 commented on clang/test/DebugInfo/Generic/preferred_name-chain.cpp:
```diff
@@ -48,7 +48,7 @@ int main() {
 // COMMON:      ![[PARAM]] = !{![[TEMPL_TYPE_PARAM:[0-9]+]]}
 // GDB:         ![[TEMPL_TYPE_PARAM]] = !DITemplateTypeParameter(name: "T", type: ![[BAR_INT_TY]])
-// LLDB:        ![[TEMPL_TYPE_PARAM]] = !DITemplateTypeParameter(name: "T", type: ![[BAR_INT_TY]])
+// LLDB:        ![[TEMPL_TYPE_PARAM]] = !DITemplateTypeParameter(name: "T", type: ![[FOO_INT]])
```
This means the type would be `Foo<int>` instead of `BarInt`. That's not ideal
